### PR TITLE
Lazy load attachments in messages view, etc.

### PR DESCRIPTION
**Pull request by charlesmchen**
A slew of optimizations around attachments that, taken together, dramatically improve responsiveness and reduce memory usage.

PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m [hunk 1]
```diff
@@ -32,4 +38,33 @@ - (CGSize)ows_adjustBubbleSize:(CGSize)bubbleSize forImage:(UIImage *)image {
     return bubbleSize;
 }
 
+- (CGSize)sizeOfImageAtURL:(NSURL *_Nullable)imageURL
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: We don't want to decode the entire image when we only need its dimensions.

File: Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.m [hunk 2]
```diff
@@ -15,13 +15,12 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface TSAnimatedAdapter ()
-//<FLAnimatedImageViewDebugDelegate>
 
 @property (nonatomic, nullable) FLAnimatedImageView *cachedImageView;
-@property (nonatomic) UIImage *image;
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.m]: We shouldn't load and retain animated GIFs in memory until we're trying to present them.

File: Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.m [hunk 3]
```diff
@@ -40,16 +39,17 @@ - (instancetype)initWithAttachment:(TSAttachmentStream *)attachment incoming:(BO
         _cachedImageView = nil;
         _attachment      = attachment;
         _attachmentId    = attachment.uniqueId;
-        _image           = [attachment image];
-        _fileData        = [NSData dataWithContentsOfURL:[attachment mediaURL]];
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.m]: Not only were we retaining the loaded image, we were also retaining its contents in memory.

File: Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m [hunk 4]
```diff
@@ -29,7 +31,7 @@ @implementation TSPhotoAdapter
 
 - (instancetype)initWithAttachment:(TSAttachmentStream *)attachment incoming:(BOOL)incoming
 {
-    self = [super initWithImage:attachment.image];
+    self = [super initWithImage:nil];
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m]: Same idea; don't load images until we need them and only retain them while they're in use.

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m (hunk 1 above)
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: Premise makes sense.

When would imageURL be nil? Seems like a callers concern.

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m [hunk 5]
```diff
@@ -32,4 +38,33 @@ - (CGSize)ows_adjustBubbleSize:(CGSize)bubbleSize forImage:(UIImage *)image {
...
+{
+    if (!imageURL) {
+        return CGSizeZero;
+    }
+
+    // With CGImageSource we avoid loading the whole image into memory.
+    CGImageSourceRef source = CGImageSourceCreateWithURL((CFURLRef)imageURL, NULL);
+    if (!source) {
+        return CGSizeZero;
```
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: should this be an assertion failure?

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m [hunk 6]
```diff
@@ -32,4 +38,33 @@ - (CGSize)ows_adjustBubbleSize:(CGSize)bubbleSize forImage:(UIImage *)image {
...
+    NSDictionary *properties
+        = (__bridge_transfer NSDictionary *)CGImageSourceCopyPropertiesAtIndex(source, 0, (CFDictionaryRef)options);
+    CGSize imageSize = CGSizeZero;
+    if (properties) {
+        NSNumber *width = properties[(NSString *)kCGImagePropertyPixelWidth];
+        NSNumber *height = properties[(NSString *)kCGImagePropertyPixelHeight];
+        if (width && height) {
+            imageSize = CGSizeMake(width.floatValue, height.floatValue);
+        }
+    }
```
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: else assertion error?

File: Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.h [hunk 7]
```diff
@@ -12,10 +12,9 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface TSAnimatedAdapter : JSQMediaItem <OWSMessageEditing, OWSMessageMediaAdapter>
 
-- (instancetype)initWithAttachment:(TSAttachmentStream *)attachment incoming:(BOOL)incoming;
+@property (nonatomic) NSString *attachmentId;
```
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.h]: readonly?

File: Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m (hunk 4 above)
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m]: Is there a reason for the `initWithImage` method to exist anymore?

File: Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m [hunk 8]
```diff
@@ -106,14 +113,6 @@ - (BOOL)canPerformEditingAction:(SEL)action
 - (void)performEditingAction:(SEL)action
 {
     NSString *actionString = NSStringFromSelector(action);
-    if (!self.image) {
```
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m]: do we now need to guard more aggressively elsewhere if image data is nil?

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: Done.

File: Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m]: I've changed this to just call `init`.

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m (hunk 5 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: Done.

File: Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m (hunk 6 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/JSQMediaItem+OWS.m]: Done.

File: Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.h (hunk 7 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSAnimatedAdapter.h]: Done.

File: Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m (hunk 8 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSPhotoAdapter.m]: I've added a few more asserts and logging.

**Pull request merged**